TailLight/SetTaillightBlack.cpp

```diff
@@ -0,0 +1,263 @@
+    if (!NT_SUCCESS(status)) {
+        KdPrintEx((DPFLTR_IHVDRIVER_ID,
+            DPFLTR_ERROR_LEVEL,
+            "TailLight: 0x%x while opening the I/O target from worker thread\n",
+            status));
+    }
+
+    return status;
+}
+
+void SetBlackCompletionRoutine(
+    _In_ WDFREQUEST Request,
+    _In_ WDFIOTARGET Target,
+    _In_ PWDF_REQUEST_COMPLETION_PARAMS Params,
+    _In_ WDFCONTEXT Context)
+{
+    UNREFERENCED_PARAMETER(Target);
+    UNREFERENCED_PARAMETER(Params);
+    UNREFERENCED_PARAMETER(Context);
+
+    NTSTATUS status = STATUS_UNSUCCESSFUL;
+    status = WdfRequestGetStatus(Request);
+    KdPrint(("TailLight: %s WdfRequestSend status: 0x%x\n", __func__, status));
+
+    // One-shot and top of stack, so delete and pray.
+    WdfObjectDelete(Request);
+}
+
+VOID SetBlackWorkItem(
+    WDFWORKITEM workItem)
+    /*++
+
+    Routine Description:
+
+        Creates a WDF workitem to do the SetBlack() call after the driver
+        stack has initialized.
+
+    Arguments:
+
+        workItem - Handle to a pre-allocated WDF work item.
+    --*/
+{
+    NTSTATUS status = STATUS_UNSUCCESSFUL;
+    WDFDEVICE device = static_cast<WDFDEVICE>(WdfWorkItemGetParentObject(workItem));
+    auto workItemContext = WdfObjectGet_SET_BLACK_WORKITEM_CONTEXT(workItem);
+
+    status = SetBlackAsync(device, workItemContext->symLink);
+
+    if (NT_SUCCESS(status)) {
+        DEVICE_CONTEXT* pDeviceContext = WdfObjectGet_DEVICE_CONTEXT(device);
+        InterlockedIncrement((PLONG)(&pDeviceContext->fSetBlackSuccess));
+    }
+
+    NukeWdfHandle<WDFWORKITEM>(workItem);
+}
+
+NTSTATUS SetBlackAsync(WDFDEVICE device, 
+    CONST UNICODE_STRING& symLink) {
+
+    PAGED_CODE();
+
+    NTSTATUS        status = STATUS_UNSUCCESSFUL;
+    WDFIOTARGET     hidTarget = NULL;
+
+    DEVICE_CONTEXT* pDeviceContext = NULL;
+
+    pDeviceContext =
+        WdfObjectGet_DEVICE_CONTEXT(device);
+
+    if (pDeviceContext == NULL) {
+        return STATUS_DEVICE_NOT_READY;
+    }
+    {
+        WDF_OBJECT_ATTRIBUTES attributes; 
+        WDF_OBJECT_ATTRIBUTES_INIT(&attributes);
+
+        // Ensure freed if fails.
+        status = WdfIoTargetCreate(
+            device,
+            &attributes,
+            &hidTarget);
+
+        if (!NT_SUCCESS(status)) {
+            KdPrintEx((DPFLTR_IHVDRIVER_ID,
+                DPFLTR_ERROR_LEVEL,
+                "TailLight: 0x%x while creating I/O target from worker thread\n",
+                status));
+            return status;
+        }
+
+        status = TryToOpenIoTarget(hidTarget, symLink);
+    }
+
+    if (NT_SUCCESS(status)) {
+        
+        UNICODE_STRING theirHwId = {};
+
+        // The hardware property ID will be attached to the IO target
+        // and thus deleted at the end of the function.
+        theirHwId = GetTargetPropertyString(hidTarget,
+            DevicePropertyHardwareID);
+        if (NT_SUCCESS(status)) {
+            if (!RtlEqualUnicodeString(&pDeviceContext->HardwareId,
```

Comment:
Are you sure that HWID is the "right" identifier to use for matching here? Won't that lead to mix-ups if attaching two identical IntelliMouse's?

I would intuitively think that the PDO name or similar would be a better identifier to use, since identical mouses will still have unique PDO names.

Reply:
Thanks for changing the code to use PDO as device identifier.